pkg/bitcoin/block.go

```diff
@@ -69,3 +105,70 @@ func (bh *BlockHeader) Hash() Hash {
 	//       3. Construct the Hash instance appropriately.
 	return Hash{}
 }
+
+// Target calculates the difficulty target of a block header.
+func (bh *BlockHeader) Target() *big.Int {
+	// A serialized 80-byte block header stores the `bits` value as a 4-byte
+	// little-endian hexadecimal value in a slot including bytes 73, 74, 75, and
+	// 76. This function's input argument is expected to be a numerical
+	// representation of that 4-byte value reverted to the big-endian order.
+	// For example, if the `bits` little-endian value in the header is
+	// `0xcb04041b`, it must be reverted to the big-endian form `0x1b0404cb` and
+	// turned to a decimal number `453248203` in order to be used as this
+	// function's input.
+	//
+	// The `bits` 4-byte big-endian representation is a compact value that works
+	// like a base-256 version of scientific notation. It encodes the target
+	// exponent in the first byte and the target mantissa in the last three bytes.
+	// Referring to the previous example, if `bits = 453248203`, the hexadecimal
+	// representation is `0x1b0404cb` so the exponent is `0x1b` while the mantissa
+	// is `0x0404cb`.
+	//
+	// To extract the exponent, we need to shift right by 3 bytes (24 bits),
+	// extract the last byte of the result, and subtract 3 (because of the
+	// mantissa length):
+	// - 0x1b0404cb >>> 24 = 0x0000001b
+	// - 0x0000001b & 0xff = 0x1b
+	// - 0x1b - 3 = 24 (decimal)
+	//
+	// To extract the mantissa, we just need to take the last three bytes:
+	// - 0x1b0404cb & 0xffffff = 0x0404cb = 263371 (decimal)
+	//
+	// The final difficulty can be computed as mantissa * 256^exponent:
+	// - 263371 * 256^24 =
+	// 1653206561150525499452195696179626311675293455763937233695932416 (decimal)
```

Comment:
In `bitcoin.go` we have `readCompactSizeUint` that uses `wire.ReadVarInt` to parse bytes into `CompactSizeUint`. If we already rely on `btcsuite/btcd` for parsing, we could eliminate most of the complexity from this function and in the function body call [`difficulty.CompactToBig`](https://github.com/btcsuite/btcd/blob/f9cbff0d819c951d20b85714cf34d7f7cc0a44b7/blockchain/difficulty.go#L62).

The same comment applies to `Difficulty()` function. In the body of that function we could call [`difficulty.CalcWork`](https://github.com/btcsuite/btcd/blob/f9cbff0d819c951d20b85714cf34d7f7cc0a44b7/blockchain/difficulty.go#L142).

I would avoid creating a separate type for this "compact" representation. We already have `type CompactSizeUint uint64` in the `bitcoin` package representing [Bitcoin CompactSize unsigned integers](https://developer.bitcoin.org/reference/transactions.html#compactsize-unsigned-integers). `CompactSizeUint` is not the same as the "compact" format of storing [the target nBits](https://developer.bitcoin.org/reference/block_chain.html#target-nbits). It makes sense to have `Target()` and `Difficulty()` functions as we have now and do not allow anything about the other "compact" representation to leak out of them.

Unit tests should stay as they are now, we should have the base coverage even if we call the library function.

Reply:
I used the `CompactToBig()` function for calculating the target.
However, we cannot use `CalcWork()` to calculate difficulty. We need to calculate difficulty the difficulty the same way 
we do in the [Solidity library](https://github.com/keep-network/bitcoin-spv/blob/8f4c751cc1fdd3e1f0c7ad40f6e4d2024a65d19f/contracts/BTCUtils.sol#L339).